### DISCUSSION TITLE
SX1280. Getting RSSI on a specific frequency

### DISCUSSION THREAD

**Discussion by VSkoshchuk** [jgromes/RadioLib#1351] [Q&A] [answered]
First of all, I'd like to say thank you for this great library!

Main Hardware:
- ESP32
- SX1280

For my task, I need to build an RSSI spectrum in the range of 2400-2420MHz, which will be used for signal pre-processing. But it seems I have encountered a problem. As far as I understand, RSSI can be obtained on a specific frequency only after the data packet is received. In my case, no information is transmitted with Lora, BLE, or GFSK modulation, on the frequencies I use. So I just received 0.0 as the RSSI value.

The code I use to measure RSSI:
```

#include <RadioLib.h>

#define LORA_SCK_PIN    5
#define LORA_MISO_PIN   3
#define LORA_MOSI_PIN   6
#define LORA_CS_PIN     7
#define LORA_RST_PIN    8
#define LORA_IO1_PIN    33

// Initialize SX1280 module
SX1280 radio = new Module(LORA_CS_PIN, LORA_IO1_PIN, LORA_RST_PIN, RADIOLIB_NC);

float startFreq = 2400.0; // Start frequency in MHz
float endFreq = 2420.0;   // End frequency in MHz
float stepSize = 1.0;     // Step size in MHz

void setup() {
  Serial.begin(115200);
  Serial.println("Hello, World!");
  vTaskDelay(1000);

  pinMode(LORA_RST_PIN, OUTPUT);
  digitalWrite(LORA_RST_PIN, HIGH);
  pinMode(LORA_SCK_PIN, OUTPUT);
  pinMode(LORA_MISO_PIN, INPUT);
  pinMode(LORA_MOSI_PIN, OUTPUT);
  SPI.begin(LORA_SCK_PIN, LORA_MISO_PIN, LORA_MOSI_PIN);

  // initialize SX1280 with default settings
  Serial.print(F("[SX1280] Initializing ... "));
  int state = radio.beginGFSK(startFreq, 800, 500, 10, 16);
  if (state == RADIOLIB_ERR_NONE) {
    Serial.println(F("success!"));
  } else {
    Serial.print(F("failed, code "));
    Serial.println(state);
    while (true) { delay(10); }
  }
}

void loop() {
  // Enter receive mode
  radio.startReceive();

  for (float freq = startFreq; freq <= endFreq; freq += stepSize) {
    // Set the current frequency
    int state = radio.setFrequency(freq);
    if (state != RADIOLIB_ERR_NONE) {
      Serial.print(F("Failed to set frequency: "));
      Serial.println(state);
      continue;
    }

    float rssi = radio.getRSSI();

    // Output the frequency and RSSI value
    Serial.print("Frequency: ");
    Serial.print(freq);
    Serial.print(" MHz, RSSI: ");
    Serial.print(rssi);
    Serial.println(" dBm");
  }

  // Exit receive mode
  radio.standby();

  vTaskDelay(1000);
}

```

Result of work:
```

> Frequency: 2400.00 MHz, RSSI: 0.00 dBm
> Frequency: 2401.00 MHz, RSSI: 0.00 dBm
> Frequency: 2402.00 MHz, RSSI: 0.00 dBm
> Frequency: 2403.00 MHz, RSSI: 0.00 dBm
> Frequency: 2404.00 MHz, RSSI: 0.00 dBm
> Frequency: 2405.00 MHz, RSSI: 0.00 dBm
> Frequency: 2406.00 MHz, RSSI: 0.00 dBm
> Frequency: 2407.00 MHz, RSSI: 0.00 dBm
> Frequency: 2408.00 MHz, RSSI: 0.00 dBm
> Frequency: 2409.00 MHz, RSSI: 0.00 dBm
> Frequency: 2410.00 MHz, RSSI: 0.00 dBm
> Frequency: 2411.00 MHz, RSSI: 0.00 dBm
> Frequency: 2412.00 MHz, RSSI: 0.00 dBm
> Frequency: 2413.00 MHz, RSSI: 0.00 dBm
> Frequency: 2414.00 MHz, RSSI: 0.00 dBm
> Frequency: 2415.00 MHz, RSSI: 0.00 dBm
> Frequency: 2416.00 MHz, RSSI: 0.00 dBm
> Frequency: 2417.00 MHz, RSSI: 0.00 dBm
> Frequency: 2418.00 MHz, RSSI: 0.00 dBm
> Frequency: 2419.00 MHz, RSSI: 0.00 dBm
> Frequency: 2420.00 MHz, RSSI: 0.00 dBm

```

Does the SX1280 support measuring RSSI on an arbitrary frequency without receiving a packet?

I would be grateful for any advice or help. I really appreciate any help you can provide.

**Comment by jgromes** [comment, marked as answer]
> Does the SX1280 support measuring RSSI on an arbitrary frequency without receiving a packet?

It does, take a look at the docs available for the method [here](https://jgromes.github.io/RadioLib/class_s_x128x.html#acaee753f0736774efc25e38f58e0cb26) - it has an overload `SX128x::getRSSI(bool packet)` that allows to specify whether to read the RSSI of the last packet, or the instantaneous value.

**Reply by VSkoshchuk** [reply]
Thank you for your answer!

Hmm, looks like SX128x does not have such a method.

My changes:
float rssi = radio.getRSSI(false);

Building logs:
```
> src/main.cpp:161:36: error: no matching function for call to 'SX1280::getRSSI(bool)'
>      float rssi = radio.getRSSI(true);

```

**Reply by jgromes** [reply]
Then you should probably update - here is that method in the library master: https://github.com/jgromes/RadioLib/blob/75f087b5c2f622d13f9c45fd4a5907f72cf09349/src/modules/SX128x/SX128x.h#L787-L792

**Reply by VSkoshchuk** [reply]
Thanks! You are right, I have used the old version.

**Reply by chrisf1874** [reply]
> > Does the SX1280 support measuring RSSI on an arbitrary frequency without receiving a packet?
> 
> It does, take a look at the docs available for the method [here]
Hi @jgromes,
I'd be very grateful if you can confirm that SX1280 GetRssiInst works outside packet reception?
I can't get it to return anything other than 0xFF.
Almost certainly something I'm doing wrong ...

**Reply by jgromes** [reply]
@chrisf1874 it should work - as listed in the docs, if you call it as `radio.getRSSI(false);` it should return the current RSSI level. Though I don't remember whether the radio needs to be in receive mode by calling `radio.startReceive()` at that point.

**Reply by chrisf1874** [reply]
@jgromes Thanks. Can confirm it does work. I needed to add a few 10s us delay after putting the radio in receive mode.
